src/phoenix/server/api/routers/v1/dataset_examples.py

```diff
@@ -0,0 +1,96 @@
+from sqlalchemy import func, select
+from starlette.requests import Request
+from starlette.responses import JSONResponse, Response
+from starlette.status import HTTP_404_NOT_FOUND
+from strawberry.relay import GlobalID
+
+from phoenix.db.models import Dataset, DatasetExample, DatasetExampleRevision, DatasetVersion
+
+
+async def list_dataset_examples(request: Request) -> Response:
+    dataset_id = GlobalID.from_id(request.path_params["id"])
+    raw_version_id = request.query_params.get("version")
+    version_id = GlobalID.from_id(raw_version_id) if raw_version_id else None
+
+    if (dataset_type := dataset_id.type_name) != "Dataset":
+        return Response(
+            content=f"ID {dataset_id} refers to a {dataset_type}", status_code=HTTP_404_NOT_FOUND
+        )
+
+    if version_id and (version_type := version_id.type_name) != "DatasetVersion":
+        return Response(
+            content=f"ID {version_id} refers to a {version_type}", status_code=HTTP_404_NOT_FOUND
+        )
+
+    async with request.app.state.db() as session:
+        # Subquery to find the maximum created_at for each dataset_example_id
+        # timestamp tiebreaks are resolved by the largest id
+        partial_subquery = select(
+            func.max(DatasetExampleRevision.id).label("max_id"),
+        ).group_by(DatasetExampleRevision.dataset_example_id)
+
+        # if a version_id is provided, filter the subquery to only include revisions from that
+        # version or earlier
+        if version_id is not None:
+            # ensure that the specified version exists
+            matched_version_id = (
+                select(DatasetVersion.id)
+                .where(DatasetVersion.id == int(version_id.node_id))
+                .scalar_subquery()
+            )
+            partial_subquery = partial_subquery.filter(
+                DatasetExampleRevision.dataset_version_id <= matched_version_id
+            )
+
+        subquery = partial_subquery.subquery()
```

Comment:
Naming feels slightly strange because we are not adding anything to the `partial_subquery` before using `subquery`.

Reply:
it's originally a partial subquery because it might be extended if a version id is passed